module/state_synchronization/indexer/execution.go

```diff
@@ -0,0 +1,207 @@
+)
+
+// ExecutionState indexes the execution state.
+type ExecutionState struct {
+	registers  storage.RegisterIndex
+	headers    storage.Headers
+	events     storage.Events
+	indexRange *SequentialIndexRange
+	log        zerolog.Logger
+}
+
+// New execution state indexer used to ingest block execution data and index it by height.
+// The passed RegisterIndex storage must be populated to include the first and last height otherwise the indexer
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*ExecutionState, error) {
+	// we must have the first height set in the storage otherwise error out
+	first, err := registers.FirstHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing first height")
+	}
+
+	// we must have the latest height set in the storage otherwise error out
+	last, err := registers.LatestHeight()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to initialize the indexer with missing latest height")
+	}
+
+	indexRange, err := NewSequentialIndexRange(first, last)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debug().
+		Uint64("first_height", first).
+		Uint64("latest_height", last).
+		Msgf("initialized indexer range")
+
+	return &ExecutionState{
+		registers:  registers,
+		headers:    headers,
+		events:     events,
+		indexRange: indexRange,
+		log:        log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+func (i *ExecutionState) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	_, err := i.indexRange.Contained(height)
+	if err != nil {
+		return nil, err
+	}
+
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - ErrIndexValue if height was not incremented in sequence
+// - storage.ErrNotFound if the block for execution data was not found
+func (i *ExecutionState) IndexBlockData(ctx context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
```

Comment:
Should we use the `irrecoverable.SignalerContext` to throw an error here?

Reply:
this error looks fine to me